core/src/main/java/org/apache/iceberg/io/inmemory/InMemoryFileStore.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io.inmemory;
+
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.ByteBuffers;
+
+/** An in-memory collection based storage for file contents keyed by a string location. */
+class InMemoryFileStore {
+
+  private final ConcurrentMap<String, byte[]> store;
+
+  InMemoryFileStore() {
+    this.store = Maps.newConcurrentMap();
+  }
+
+  /** Put the file contents at the given location, overwrite if it already exists. */
+  public void put(String location, ByteBuffer data) {
+    // Copy the contents and store it.
+    store.put(location, ByteBuffers.toByteArray(data));
+  }
+
+  /** Get the file contents for the given location. */
+  public Optional<ByteBuffer> get(String location) {
+    return Optional.ofNullable(store.get(location))
+        .map(
+            bytes -> {
```

Comment:
You can store the `ByteBuffer` in `put()` using `ByteBuffer.asReadOnlyBuffer()` and then return just return `ByteBuffer.duplicate()` here instead of the byte[] copy + wrapping.

Reply:
good point, done